rubicon/objc/async.py

```diff
@@ -0,0 +1,583 @@
+            libcf.CFRunLoopRemoveSource(
+                self._loop._cfrunloop, self._src, kCFRunLoopCommonModes
+            )
+            return
+
+        if callbackType == kCFSocketReadCallBack:
+            callback, args = self._reader
+        elif callbackType == kCFSocketWriteCallBack:
+            callback, args = self._writer
+        else:
+            callback = None
+
+        if callback:
+            callback(*args)
+
+    def __init__(self, *, loop, fd):
+        """
+        Register a file descriptor with the C{CFRunLoop}, or modify its state
+        so that it's listening for both notifications (read and write) rather
+        than just one; used to implement C{addReader} and C{addWriter}.
+        @param fd: The file descriptor.
+        @type fd: L{int}
+        @param descr: the L{IReadDescriptor} or L{IWriteDescriptor}
+        @param flag: the flag to register for callbacks on, either
+            C{kCFSocketReadCallBack} or C{kCFSocketWriteCallBack}
+        """
+        super().__init__(CFSocketCallback(self._cf_socket_callback), None, loop)
+
+        # Retain a reference to the Handle
+        self._loop._sockets[fd] = self
+        self._reader = None
+        self._writer = None
+
+        self._fd = fd
+        self._cf_socket = libcf.CFSocketCreateWithNative(
+            kCFAllocatorDefault, self._fd,
+            kCFSocketReadCallBack | kCFSocketWriteCallBack |
+            kCFSocketConnectCallBack,
+            self._callback,
+            None
+        )
+        libcf.CFSocketSetSocketFlags(
+            self._cf_socket,
+            kCFSocketAutomaticallyReenableReadCallBack |
+            kCFSocketAutomaticallyReenableWriteCallBack
+
+            # # This extra flag is to ensure that CF doesn't (destructively,
+            # # because destructively is the only way to do it) retrieve
+            # # SO_ERROR
+            # 1 << 6
+        )
+        self._src = libcf.CFSocketCreateRunLoopSource(kCFAllocatorDefault, self._cf_socket, 0)
+        libcf.CFRunLoopAddSource(self._loop._cfrunloop, self._src, kCFRunLoopCommonModes)
+        libcf.CFSocketDisableCallBacks(
+            self._cf_socket,
+            kCFSocketReadCallBack | kCFSocketWriteCallBack |
+            kCFSocketConnectCallBack
+        )
+
+    def enable_read(self, callback, args):
+        libcf.CFSocketEnableCallBacks(self._cf_socket, kCFSocketReadCallBack)
+        self._reader = (callback, args)
+
+    def disable_read(self):
+        libcf.CFSocketDisableCallBacks(self._cf_socket, kCFSocketReadCallBack)
+        self._reader = None
+        self.cancel()
+
+    def enable_write(self, callback, args):
+        libcf.CFSocketEnableCallBacks(self._cf_socket, kCFSocketWriteCallBack)
+        self._writer = (callback, args)
+
+    def disable_write(self):
+        libcf.CFSocketDisableCallBacks(self._cf_socket, kCFSocketWriteCallBack)
+        self._writer = None
+        self.cancel()
+
+    def cancel(self):
+        """
+        """
```

Comment:
This should probably be removed (or filled).

Reply:
Unfortunately, this comment has been cancelled. :-)